src/ForwardedHttpMiddleware.cs

```diff
@@ -1,27 +1,885 @@
+            {
+                feature.OriginalRemoteIpAddress = connection.RemoteIpAddress;
+            }
+
+            feature.OriginalRemotePort = connection.RemotePort;
+
+            var request = context.Request;
+
+            feature.OriginalProto = request.Scheme;
+
+            feature.OriginalHost = request.Host.Value;
+
+            context.Features.Set<IForwardedHttpFeature>(feature);
+        }
+
+        private static (StringSegment ip, StringSegment port) GetIpPort(StringSegment ipAndPortText)
+        {
+            StringSegment ip;
+            StringSegment port;
+
+
+            if (ipAndPortText.StartsWith("[", StringComparison.OrdinalIgnoreCase))
+            {
+                // ipv6 
+
+                if (ipAndPortText.EndsWith("]", StringComparison.OrdinalIgnoreCase))
+                {
+                    // ipv6 without port
+                    // [2001:db8:cafe::17]
+
+                    ip = ipAndPortText;
+                    port = StringSegment.Empty;
+                }
+                else
+                {
+                    // ipv6 with port
+                    // [2001:db8:cafe::17]:12345
+
+                    var index = ipAndPortText.IndexOf(']');
+                    ip = ipAndPortText.Subsegment(1, index - 1);
+                    port = ipAndPortText.Subsegment(index + 2);
+                }
+            }
+            else
+            {
+                // ipv4 
+
+                var colonIndex = ipAndPortText.IndexOf(':');
+                var hasPort = colonIndex > 0;
+                if (hasPort)
+                {
+                    // ipv4 with port
+                    // 192.0.2.43:47011
+
+                    ip = ipAndPortText.Subsegment(0, colonIndex);
+                    port = ipAndPortText.Subsegment(colonIndex + 1);
+                }
+                else
+                {
+                    // ipv4 without port
+                    // 192.0.2.43
+
+                    ip = ipAndPortText;
+                    port = StringSegment.Empty;
+                }
+            }
+
+            return (ip, port);
+        }
+
+        private bool IsValidObfuscatedNode(StringSegment ipAndPortText)
+        {
+            if (ipAndPortText[0] != '_')
+            {
+                return false;
+            }
+
+            for (var i = 1; i < ipAndPortText.Length; i++)
+            {
+                if (!IsValidObfuscatedChar(ipAndPortText[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private bool IsUnknownNode(StringSegment remoteIpAndPortText)
+        {
+            // Need any string comparision?
+            return remoteIpAndPortText.Equals("unknown");
+        }
+
+        // Empty was checked for by the caller
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryValidateScheme(StringSegment scheme)
```

Comment:
I recommend moving all of this validation code to another class for both organization and testing.